src/System.Private.CoreLib/shared/System/Threading/WaitHandle.cs

```diff
@@ -0,0 +1,449 @@
+        {
+            var timeoutMilliseconds = (long)timeout.TotalMilliseconds;
+            if (timeoutMilliseconds < -1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(timeout), SR.ArgumentOutOfRange_NeedNonNegOrNegative1);
+            }
+            if (timeoutMilliseconds > int.MaxValue)
+            {
+                throw new ArgumentOutOfRangeException(nameof(timeout), SR.ArgumentOutOfRange_LessEqualToIntegerMaxVal);
+            }
+            return (int)timeoutMilliseconds;
+        }
+
+        public virtual void Close() => Dispose();
+
+        protected virtual void Dispose(bool explicitDisposing)
+        {
+            _waitHandle?.Close();
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        public virtual bool WaitOne(int millisecondsTimeout)
+        {
+            if (millisecondsTimeout < -1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(millisecondsTimeout), SR.ArgumentOutOfRange_NeedNonNegOrNegative1);
+            }
+
+            return WaitOneNoCheck(millisecondsTimeout);
+        }
+
+        private bool WaitOneNoCheck(int millisecondsTimeout)
+        {
+            Debug.Assert(millisecondsTimeout >= -1);
+
+            // The field value is modifiable via the public <see cref="WaitHandle.SafeWaitHandle"/> property, save it locally
+            // to ensure that one instance is used in all places in this method
+            SafeWaitHandle waitHandle = _waitHandle;
+            if (waitHandle == null)
+            {
+                // Throw ObjectDisposedException for backward compatibility even though it is not be representative of the issue
+                throw new ObjectDisposedException(null, SR.ObjectDisposed_Generic);
+            }
+
+            bool success = false;
+            try
+            {
+                int waitResult;
+
+                waitHandle.DangerousAddRef(ref success);
+
+                SynchronizationContext context = SynchronizationContext.Current;
+                if (context != null && context.IsWaitNotificationRequired())
+                {
+                    waitResult = context.Wait(new[] { waitHandle.DangerousGetHandle() }, false, millisecondsTimeout);
+                }
+                else
+                {
+                    waitResult = WaitOneCore(waitHandle.DangerousGetHandle(), millisecondsTimeout);
+                }
+
+                if (waitResult == WaitAbandoned)
+                {
+                    throw new AbandonedMutexException();
+                }
+
+                return waitResult != WaitTimeout;
+            }
+            finally
+            {
+                if (success)
+                    waitHandle.DangerousRelease();
+            }
+        }
+
+        private static void EnsureRentalItemsInitialized()
+        {
+            if (!_rentalItemsInitialized)
```

Comment:
This `bool` flag and the `EnsureRentalItemsInitialized` method should not be needed.

Reply:
Ah, the flag is needed - to avoid creating Gen2 garbage on recursive waits...

Reply:
Maybe worth a comment.

Reply:
The waits can be re-entrant, so I cannot just check for `_safeWaitHandlesForRent == null` because that will happen when a) the array was not initialised b) a re-entrant wait is happening. I can get rid of it at the cost of always allocating of 64 entries for the re-entrant waits and then returning the array from the inner-most wait.

Reply:
We can always return the array from the outer most wait, overwrite whatever is there.

The flag should not be needed if we do that.